controllers/azuremachine_reconciler.go

```diff
@@ -69,253 +58,62 @@ func newAzureMachineService(machineScope *scope.MachineScope, clusterScope *scop
+	if err := s.networkInterfacesSvc.Reconcile(ctx); err != nil {
+		return errors.Wrap(err, "failed to create network interface")
 	}
 
-	vm, vmErr := s.reconcileVirtualMachine(ctx, azure.GenerateNICName(s.machineScope.Name()))
-	if vmErr != nil {
-		return nil, errors.Wrapf(vmErr, "failed to create VM %s ", s.machineScope.Name())
+	if err := s.virtualMachinesSvc.Reconcile(ctx); err != nil {
+		return errors.Wrap(err, "failed to create virtual machine")
 	}
 
-	err = s.roleAssignmentsSvc.Reconcile(ctx)
-	if err != nil {
-		return nil, errors.Wrap(err, "unable to create role assignment")
+	if err := s.roleAssignmentsSvc.Reconcile(ctx); err != nil {
+		return errors.Wrap(err, "unable to create role assignment")
 	}
 
-	return vm, nil
+	return nil
 }
 
 // Delete deletes all the services in pre determined order
 func (s *azureMachineService) Delete(ctx context.Context) error {
-	vmSpec := &virtualmachines.Spec{
-		Name: s.machineScope.Name(),
-	}
-
-	err := s.virtualMachinesSvc.Delete(ctx, vmSpec)
-	if err != nil {
+	if err := s.virtualMachinesSvc.Delete(ctx); err != nil {
 		return errors.Wrapf(err, "failed to delete machine")
 	}
 
-	err = s.networkInterfacesSvc.Delete(ctx)
-	if err != nil {
-		return errors.Wrapf(err, "Unable to delete network interface")
+	if err := s.networkInterfacesSvc.Delete(ctx); err != nil {
+		return errors.Wrapf(err, "failed to delete network interface")
 	}
 
-	err = s.inboundNatRulesSvc.Delete(ctx)
-	if err != nil {
-		return errors.Wrapf(err, "Unable to delete inbound NAT rule")
+	if err := s.inboundNatRulesSvc.Delete(ctx); err != nil {
+		return errors.Wrapf(err, "failed to delete inbound NAT rule")
 	}
 
-	err = s.publicIPsSvc.Delete(ctx)
-	if err != nil {
+	if err := s.publicIPsSvc.Delete(ctx); err != nil {
 		return errors.Wrap(err, "failed to delete public IPs")
 	}
 
-	err = s.disksSvc.Delete(ctx)
-	if err != nil {
-		return errors.Wrapf(err, "Failed to delete OS disk of machine %s", s.machineScope.Name())
+	if err := s.disksSvc.Delete(ctx); err != nil {
+		return errors.Wrap(err, "failed to delete OS disk")
 	}
 
 	return nil
 }
 
-func (s *azureMachineService) VMIfExists(ctx context.Context, id *string) (*infrav1.VM, error) {
-	if id == nil {
-		s.clusterScope.Info("VM does not have an ID")
-		s.clusterScope.Info("VM does not have an ID")
-		return nil, nil
-	}
-
-	vmSpec := &virtualmachines.Spec{
-		Name: s.machineScope.Name(),
-	}
-	vm, err := s.virtualMachinesSvc.Get(ctx, vmSpec)
-
-	if azure.ResourceNotFound(err) {
-		return nil, nil
-	}
-
-	if err != nil {
-		return nil, errors.Wrap(err, "Failed to get VM")
-	}
-
-	klog.V(2).Infof("Found VM for AzureMachine %s", s.machineScope.Name())
-
-	return vm, nil
-}
-
-// getVirtualMachineZone gets a random availability zones from available set,
-// this will hopefully be an input from upstream machinesets so all the vms are balanced
-func (s *azureMachineService) getVirtualMachineZone(ctx context.Context) (string, error) {
```

Comment:
So I thought it was okay to remove this and that it would automatically get taken care of by failure domains, but cluster api actually doesn't spread Machines on failure domains that are part of a machineDeployment/machineSet (only control plane machines part of KubeadmControlPlane). There's actually an open issue to potentially enable machineSets to spread machines over failure domains https://github.com/kubernetes-sigs/cluster-api/issues/3358

So essentially by removing this, it would cause worker node VMs to not be assigned a Zone until #3358 is implemented. This is potentially a regression. Reasons I'm considering accepting that temporary regression right now:
1) this doesn't actually spread VMs on zones, it picks a "random" zone (not that random it just picks the first one, see line 193).
2) The right way to spread machines part of a MachineSet on zones is with https://github.com/kubernetes-sigs/cluster-api/issues/3358, not having this hacky workaround might be a forcing function to actually do it right.
3) There is another workaround apparently which is to create one machine deployment per failure domain and scale each one independently (documented in https://github.com/kubernetes-sigs/cluster-api/issues/3358)
4) Doing this random AZ assignment at the VM level means the failureDomain on the machine spec is not set, which might cause friction with the cluster api cluster-autoscaler implementation and also means this wouldn't be compatible with a future implementation of failure domains on MachineSet (ie. existing VMs would break as we try to mutate the zone)
5) There is an in-progress MachinePool failure domains implementation which might be a good enough solution for worker nodes having AZs for now (ie. "if you need AZs, use vmss or use several deployments).

Thoughts? Happy to talk through this, it's a lot of writing :)

Reply:
filed https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/844 to clarify docs

Reply:
I vote to remove `getVirtualMachineZone`. The non-deterministic behavior of assigning zones is unexpected from the users' point of view and can cause behavioral changes amongst VMs in a machine deployment.